libtock/alarm_timer.c

```diff
@@ -4,11 +4,131 @@
 #include <limits.h>
 #include <stdlib.h>
 
+#define MAX_TICKS UINT32_MAX
+
 // Returns 0 if a <= b < c, 1 otherwise
 static int within_range(uint32_t a, uint32_t b, uint32_t c) {
   return (b - a) < (b - c);
 }
 
+static uint32_t ticks_to_ms(uint32_t ticks) {
+  // `ticks_to_ms`'s conversion will be accurate to within the range
+  // 0 to 1 milliseconds less than the exact conversion
+  // (true millisecond conversion - [0,1) milliseconds).
+
+  uint32_t frequency;
+  alarm_internal_frequency(&frequency);
+
+  uint32_t seconds = (ticks / frequency);
+  uint32_t milliseconds_per_second = 1000;
+
+  // Calculate the conversion of full seconds to ticks.
+  uint32_t milliseconds = seconds * milliseconds_per_second;
+
+  // To get conversion accuracy within 1 millisecond, the conversion
+  // must also convert partial seconds.
+  uint32_t leftover_ticks = ticks % frequency;
+
+  // This calculation is mathematically equivalent to doing:
+  //
+  //   `leftover_ticks` / (`frequency` / `milliseconds_per_second`)
+  //
+  // This is taking the remaining unconverted ticks (leftover_ticks)
+  // and dividing by the number of ticks per millisecond
+  // (`frequency` (ticks per second) / `1000` milliseconds per second)
+  // The division is done this way because of the same argument in
+  // `ms_to_ticks`.
+  milliseconds += (leftover_ticks * milliseconds_per_second) / frequency;
+  return milliseconds;
+}
+
+// WARNING: This function can silently overflow if the output
+// number of ticks is larger than `UINT32_MAX`.
```

Comment:
This does not seem ideal considering that libtock-c assumes:

```
 * The client should not assume anything about the underlying clock used by an
 * implementation other than that it is running at sufficient frequency to
 * deliver at least millisecond granularity and that it is a 32-bit clock (i.e.
 * it will wrap at 2^32 clock ticks).
 ```

Perhaps it would be better to pass a `uint32_t` reference that this function updates and instead returns a returncode?